### PR TITLE
remove SOC includes SOAD attachment validation check

### PR DESCRIPTION
**Pull request by CHssmith2**
1. Removed the validation check to ensure a statement-of-affairs-director attachment is present with the filing of a statement-of-concurrence.
2. Added a unit test to ensure that when a statement-of-concurrence is filed separately then it still requires the presence of at least one practitioner.

### CODE REVIEW

**Comment by nevans4-ch** [comment]
Unit tests are currently failing with:

  * /tmp/build/9780d8b7/go/src/github.com/companieshouse/insolvency-api/service/insolvency_service_test.go 
  Line 286:
  Expected collection to have length equal to [1], but its length was [0] instead! contents: &[]

(I think just means you need to take out the check relating to the removed validation.)

I'd also expected to see a change in handlers/soa_resource.go to add statement-of-concurrence into the allowed attachment types to allow a date to be associated with it?

File: handlers/soa_resource.go [hunk 1]
```diff
@@ -64,9 +64,9 @@ func HandleCreateStatementOfAffairs(svc dao.Service, helperService utils.HelperS
 		}
 
 		// Validate the supplied attachment is a valid type
-		if attachment.Type != "statement-of-affairs-director" && attachment.Type != "statement-of-affairs-liquidator" {
+		if attachment.Type != "statement-of-affairs-director" && attachment.Type != "statement-of-affairs-liquidator"  && attachment.Type != "statement-of-concurrence" {
```
Comment (pparlett) [inline handlers/soa_resource.go:67]: in this modified code you have not used the constants for "statement-of-affairs-director"?

File: handlers/soa_resource.go [hunk 2]
```diff
@@ -64,9 +64,9 @@ func HandleCreateStatementOfAffairs(svc dao.Service, helperService utils.HelperS
...
 			err := fmt.Errorf("attachment id [%s] is an invalid type for this request: %v", statementDao.Attachments[0], attachment.Type)
-			responseMessage := "attachment is not a statement-of-affairs-director or statement-of-affairs-liquidator"
+			responseMessage := "attachment is not a statement-of-affairs-director, statement-of-affairs-liquidator or a statement-of-concurrence"
```
Comment (pparlett) [inline handlers/soa_resource.go:69]: ditto previous

File: service/insolvency_service.go [hunk 3]
```diff
@@ -123,16 +111,17 @@ func ValidateInsolvencyDetails(insolvencyResource models.InsolvencyResourceDao)
...
-	if (hasStatementOfAffairsDirector || hasStateOfAffairsLiquidator) && (insolvencyResource.Data.StatementOfAffairs == nil || insolvencyResource.Data.StatementOfAffairs.StatementDate == "") {
-		validationError := fmt.Sprintf("error - a date of statement of affairs must be present as there is an attachment with type [%s] or [%s] for insolvency case with transaction id [%s]", constants.StatementOfAffairsDirector.String(), constants.StatementOfAffairsLiquidator.String(), insolvencyResource.TransactionID)
+	if (hasStatementOfAffairsDirector || hasStateOfAffairsLiquidator || hasStatementOfConcurrence) && (insolvencyResource.Data.StatementOfAffairs == nil || insolvencyResource.Data.StatementOfAffairs.StatementDate == "") {
+		validationError := fmt.Sprintf("error - a date of statement of affairs must be present as there is an attachment with a type of [%s], [%s], or a [%s] for insolvency case with transaction id [%s]", constants.StatementOfAffairsDirector.String(), constants.StatementOfConcurrence.String(), constants.StatementOfAffairsLiquidator.String(), insolvencyResource.TransactionID)
```
Comment (pparlett) [inline service/insolvency_service.go:117]: this function is long partly due to many occurrences of, considering refactoring into a function:
		validationError := ...
		log.Info(validationError)
		validationErrors = addValidationError(validationErrors, validationError, "attachment ids do not match")

File: service/insolvency_service.go [hunk 4]
```diff
@@ -123,16 +111,17 @@ func ValidateInsolvencyDetails(insolvencyResource models.InsolvencyResourceDao)
 	}
 
 	// Check if "statement-of-affairs-director" or "statement-of-affairs-liquidator" has been filed, if so, then a statement date must be present
```
Comment (nevans4-ch) [inline service/insolvency_service.go:113]: Would be good to edit this comment to include the SoC

**Review by pparlett — CHANGES_REQUESTED** [review]
Code logic looks OK as following existing pattern
Code builds and unit tests run OK, new/modified code has full coverage

There are two SonarQube complexity warnings on the modified functions in soa_resource and insolvency_service. The option here is to break these complex functions down into (clearly named) sub-functions, use existing comments for ideas. This can be done automatically using the VSCode refactoring tool. There is good unit test coverage so this should be safe to do. Ask Gomathi if she wants you to do this refactoring.

General comment, the go code for enums looks "horrible", is there no way of making constants.StatementOfAffairsLiquidator.String() shorter?, in Java it is a simple enum value :-( It seems because they are are horrible they are not properly used in ALL of the code within this program, e.g. do a file search for "statement-of-affairs-director", this should be refactored.

[So much validation code, this would be a suitable for application of a rules engine.]

File: handlers/soa_resource.go (hunk 1 above)
Comment (pparlett) [reply handlers/soa_resource.go]: done

**Review by pparlett — APPROVED** [review]
The two SonarQube warnings have been discussed and it has been identified that although they relate to "New Code" the warnings are pre-existing (also present on the main branch). Refactoring to remove the warnings is too much extra work for this ticket.

**Pull request merged**